README typo causes app crash

The app crashes if Go templates are not well-defined.

Steps to reproduce:

1. Copy this line: https://github.com/roboll/helmfile/blame/master/README.md#L405
2. Run any `helmfile` command (I did `lint` and `sync`)
3. See a beautiful crash stacktrace:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x136ac2e]

goroutine 1 [running]:
github.com/roboll/helmfile/vendor/github.com/urfave/cli.HandleAction.func1(0xc42014ef08)
	/go/src/github.com/roboll/helmfile/vendor/github.com/urfave/cli/app.go:474 +0x279
panic(0x13d7220, 0x1683ab0)
	/usr/local/go/src/runtime/panic.go:502 +0x229
bytes.(*Buffer).Bytes(...)
	/usr/local/go/src/bytes/buffer.go:55
main.(*twoPassRenderer).renderEnvironment(0xc42014e8c0, 0xc4200c2900, 0x219, 0x419, 0x104f210, 0xc42014e640, 0x10988bb)
	/go/src/github.com/roboll/helmfile/main.go:657 +0x1ce
main.(*twoPassRenderer).renderTemplate(0xc42014e8c0, 0xc4200c2900, 0x219, 0x419, 0x419, 0x0, 0x0)
	/go/src/github.com/roboll/helmfile/main.go:673 +0x6a
main.(*app).FindAndIterateOverDesiredStates(0xc42014ea10, 0x7ffeefbfed60, 0x5b, 0xc420101220, 0x0, 0x0, 0x16b0280, 0x0, 0x0, 0x7ffeefbfed57, ...)
	/go/src/github.com/roboll/helmfile/main.go:711 +0x2cd
main.findAndIterateOverDesiredStatesUsingFlagsWithReverse(0xc4200d2c80, 0x0, 0xc4200714d0, 0x1, 0xc4200714d0)
	/go/src/github.com/roboll/helmfile/main.go:604 +0x387
main.findAndIterateOverDesiredStatesUsingFlags(0xc4200d2c80, 0xc4200714d0, 0x20, 0xc420099a10)
	/go/src/github.com/roboll/helmfile/main.go:568 +0x3a
main.main.func5(0xc4200d2c80, 0x0, 0x0)
	/go/src/github.com/roboll/helmfile/main.go:240 +0x65
reflect.Value.call(0x13bd020, 0x1460bc0, 0x13, 0x144163d, 0x4, 0xc42014eea8, 0x1, 0x1, 0xc420030080, 0x1434ec0, ...)
	/usr/local/go/src/reflect/value.go:447 +0x969
reflect.Value.Call(0x13bd020, 0x1460bc0, 0x13, 0xc42014eea8, 0x1, 0x1, 0x4, 0xc4200a1350, 0x5)
	/usr/local/go/src/reflect/value.go:308 +0xa4
github.com/roboll/helmfile/vendor/github.com/urfave/cli.HandleAction(0x13bd020, 0x1460bc0, 0xc4200d2c80, 0x0, 0x0)
	/go/src/github.com/roboll/helmfile/vendor/github.com/urfave/cli/app.go:483 +0x173
github.com/roboll/helmfile/vendor/github.com/urfave/cli.Command.Run(0x1441a25, 0x4, 0x0, 0x0, 0x0, 0x0, 0x0, 0x145440d, 0x27, 0x0, ...)
	/go/src/github.com/roboll/helmfile/vendor/github.com/urfave/cli/command.go:186 +0x999
github.com/roboll/helmfile/vendor/github.com/urfave/cli.(*App).Run(0xc420001200, 0xc42001c0c0, 0x6, 0x6, 0x0, 0x0)
	/go/src/github.com/roboll/helmfile/vendor/github.com/urfave/cli/app.go:237 +0x546
main.main()
	/go/src/github.com/roboll/helmfile/main.go:503 +0x1ffc
```

Firstly, the README should be updated because an `if` statement needs no parentheses in Go templates: https://gist.github.com/techslides/8957209

Secondly, exceptions should be gracefully handled with a nice error message instead of crashing the app.

Used with helmfile version 0.37

--- a/README.md
+++ b/README.md
@@ -402,7 +402,7 @@ environments:
 
 releases:
 
-{{ if (eq .Environment.Name "production" }}
+{{ if eq .Environment.Name "production" }}
 - name: newrelic-agent
   # snip
 {{ end }}

--- a/main.go
+++ b/main.go
@@ -3,16 +3,15 @@ package main
 import (
 	"bytes"
 	"fmt"
+	"io/ioutil"
 	"log"
 	"os"
+	"os/exec"
 	"os/signal"
 	"path/filepath"
 	"sort"
-	"syscall"
-
-	"io/ioutil"
-
 	"strings"
+	"syscall"
 
 	"github.com/roboll/helmfile/args"
 	"github.com/roboll/helmfile/environment"
@@ -22,7 +21,6 @@ import (
 	"github.com/urfave/cli"
 	"go.uber.org/zap"
 	"go.uber.org/zap/zapcore"
-	"os/exec"
 )
 
 const (
@@ -726,8 +724,12 @@ func (r *twoPassRenderer) renderEnvironment(content []byte) environment.Environm
 
 	// parse as much as we can, tolerate errors, this is a preparse
 	yamlBuf, err := firstPassRenderer.RenderTemplateContentToBuffer(content)
-	if err != nil && logger != nil {
+	if err != nil && r.logger != nil {
 		r.logger.Debugf("first-pass rendering input of \"%s\":\n%s", r.filename, prependLineNumbers(string(content)))
+		if yamlBuf == nil { // we have a template syntax error, let the second parse report
+			r.logger.Debugf("template syntax error: %v", err)
+			return firstPassEnv
+		}
 	}
 	c := state.NewCreator(r.logger, r.reader, r.abs)
 	c.Strict = false

--- a/main_test.go
+++ b/main_test.go
@@ -2,10 +2,12 @@ package main
 
 import (
 	"fmt"
+	"os"
 	"path/filepath"
 	"strings"
 	"testing"
 
+	"github.com/roboll/helmfile/helmexec"
 	"github.com/roboll/helmfile/state"
 	"gopkg.in/yaml.v2"
 )
@@ -16,7 +18,7 @@ func makeRenderer(readFile func(string) ([]byte, error), env string) *twoPassRen
 		env:       env,
 		namespace: "namespace",
 		filename:  "",
-		logger:    logger,
+		logger:    helmexec.NewLogger(os.Stdout, "debug"),
 		abs:       filepath.Abs,
 	}
 }
@@ -226,5 +228,29 @@ func TestReadFromYaml_RenderTemplateWithNamespace(t *testing.T) {
 
 	if state.Releases[0].Name != "namespace-myrelease" {
 		t.Errorf("release name should be namespace-myrelease")
+	}
+}
+
+func TestReadFromYaml_HelfileShouldBeResilentToTemplateErrors(t *testing.T) {
+	yamlContent := []byte(`environments:
+  staging:
+	production:
+
+releases:
+{{ if (eq .Environment.Name "production" }}  # notice syntax error: unclosed left paren
+- name: prod-myrelease
+{{ else }}
+- name: myapp
+{{ end }}
+  chart: mychart
+`)
+	fileReader := func(filename string) ([]byte, error) {
+		return yamlContent, nil
+	}
+
+	r := makeRenderer(fileReader, "staging")
+	_, err := r.renderTemplate(yamlContent)
+	if err == nil {
+		t.Fatalf("wanted error, none returned")
 	}
 }

